### PR TITLE
[SPARK-42651][SQL] Optimize global sort to driver sort

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

This pr adds a new physical plan `DriverSortExec`. It represents a special case of global `SortExec`  which is the root node. Then we can save one shuffle.

Add a new config `spark.sql.execution.driverSortThreshold` to control if it's safe to do sort at driver side. We should make sure the max rows of plan is small enough to avoid oom and preformance regression.

This optimization should work fine since It gets benefits from AQE framework that the max rows of logical plan is accurate and it can be propagated cross most of nodes.

### Why are the changes needed?

Improve performance

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

add test and do some benchmark

Sort 50, 000 rows for such case `df.sort.collect`:
```
OpenJDK 64-Bit Server VM 1.8.0_362-b09 on Linux 5.15.0-1033-azure
Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
DriverSort:                               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Sort                                               1158           1355         210          0.0       23169.3       1.0X
DriverSort                                          636            680          55          0.1       12715.2       1.8X
```

### CODE REVIEW

**Comment by zhengruifeng** [comment]
just a idea, what about: making  `SortExec` requires `SinglePartition`

**Comment by ulysses-you** [comment]
@zhengruifeng thank you for your thought.
 
The original idea of driver sort is to avoid one shuffle. Requires SinglePartition seems does not help since it still requires a shuffle.

Besides, finally, the result would go to driver, i.e. `df.sort.collect` (It's the reason I match `ReturnAnswer`), so it should be fine to do at driver. Plus, do sort at driver is not the first code place except driver sort. e.g., the merge function of rdd.takeOrdered. It should be safe if the size of plan is small enough.

**Comment by ulysses-you** [comment]
cc @cloud-fan @viirya thank you

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, NamedExpression, SortOrder, UnsafeProjection}
+import org.apache.spark.sql.catalyst.expressions.codegen.LazilyGeneratedOrdering
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, SinglePartition}
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+
+/**
+ * A special operator for global [[SortExec]] if it is the root node with/without [[ProjectExec]].
+ */
+case class DriverSortExec(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala:32]: Can we use `TakeOrderedAndProjectExec` to replace it? They seem to be similar. What if we give a large enough limit to `TakeOrderedAndProjectExec`?

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala:32]: We have a rule `EliminateLimits` to remove limit if it's large enough, i.e., the actually row count is less than or equal to limit number. So if the `Sort` is the root node that means:
1. the user does not specify the limit
2. the limit has no meaning and has been eliminated

I think if the data size is small enough, `DriverSortExec` is faster than `TakeOrderedAndProjectExec`:
- `DriverSortExec` only sort data once at driver side
- `TakeOrderedAndProjectExec` sort the data at executor side then merge the sorted data at driver side

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala:32]: `DriverSortExec` must has a local sort below, isn't it?

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala:32]: no, it is the local sort..

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala:32]: have you tried to do a local sort first? I think merging sorted data in the driver side should be faster.

**Comment by mridulm** [comment]
Meta comment: moving this to the driver has potential for destabilizing the application.
In SPARK-36419, we added option to move the final `treeAggregate` to executor given the scale challenges seen.

One option would be to do this in an executor as @zhengruifeng suggested, and that should be much more scalable as a solution.

**Pull request closed without merging**